README.txt
=== WP Custom Fields Search ===
Contributors: don@webhammer.co.uk
Donate link: https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=don@webhammer.co.uk&item_name=Custom Search Donation&currency_code=GBP
Tags: search,custom fields,widget,sidebar
Requires at least: 2.5
Tested up to: 4.2.3
Stable tag: 0.3.27

This plugin allows multiple form inputs to be configured to search different aspects of a post including custom fields.


== Description ==

This is a search plugin for wordpress, designed to filter posts in a more structured way than the default wordpress search. Specifically it allows multiple form inputs to be configured to search different aspects of a post, i.e. one term could search the post content, one the post title and one could search a custom field associated with the post.

The blog admin is able to build a customised search form. It allows you to search by the standard post information and by custom fields and to choose what HTML form elements should be presented for each search field.

I originally developed this plugin for a client project. Basically the problem they had was that they were storing real estate data as wordpress posts with associated custom fields and they wanted to search for entries based on price, location and features which was not possible with the default wordpress search.

The plugin is undergoing a period of change, there will be a full rebuild in the near future and there will be a change of format to add in premium versions of the plugin which will fund further development and support to the community using this plugin.

This final release fixes an incompatibility with recent versions of wordpress and pulls in a few of the features I have developed for bespoke clients over the last four years, some of these new features are a little rough around the edges but hopefully they will be of use to some of you.

== Installation ==
I will try to put more complete documentation together at some point but for now this is a quick start guide to building a search form.i

1. Install the plugin.  Upload the directory wp-content-search to your /wp-content/plugins/ directory.
2. Activate it through the plugins menu in Wordpress

Method 1: Add the search as a sidebar widget

1. Add the widget to the sidebar on which you want it to appear. Go to the appearance > widgets section of the admin page, in the list on the left find the widget "Custom Fields Search" and click the add button next to it. It should appear in the list on the right. Click save changes
2. You should now have a basic search form on the front of your site, to customise this further read on to see how to reconfigure your fields.

Method 2: Add the search code directly into your template

1. Go to the custom fields config page in your admin section, Settings > WP Custom Fields Search
2. Copy the example PHP code from this page into your template file.
3. You should now have a basic search form on the front of your site, to customise this further follow the instructions below for setting up fields.

Method 3: Add a tag into your posts/pages

1. Go to the custom fields config page in your admin section, Settings > WP Custom Fields Search
2. Copy the example tag for posts from this page into the content of the page or post in which you want the form to appear.
3. You should now have a basic search form in this page/post, to customise this further follow the instructions below for setting up fields.

== Freqently Asked Questions ==

= How do choose what options are shown in Drop Down lists or Radio Buttons =

Each of these widgets has a special input field called "Drop Down Options" and "Radio Button Options" respectively.  If you put a comma separated list of values in here these will be shown as the results.

If you want to show a label that is different to the search term you can separate each item with a colon ':' so if I have a category called Cat1 which has my favourite posts I could use Cat1:Favourite Posts to show one value while searching another.  This is particular useful for the any option so something like the following ":Any,1:Option 1,2,3" would produce a dropdown where the first element has the label any but has no value so will not actually restrict the search.

= Can the plugin search pages instead of posts? =

Yes, it can, although the interface for this is a bit awkard at the moment.

You will need to create a hidden field, which searches on the post_type data type, then give this field the value "Page".  This field will not show up in the interface but will affect the search results giving you page search.

= How to sort results = 

The sorting functionality is not 100% complete yet.  At present you can set the sort order of results, but it's a little counter intuitive.  Create a new field on your search form, set the data type and data field to the field you want to sort by, set the compare function to "Sort Order", and set the widget to "Hidden Constant".  To sort in reverse order set the constant value to "descending".

== Screenshots ==
1. Front end search box
2. Back end admin screen showing field config

== Setting up the fields ==

Each field has a number of settings which control the way the input appears to the user and the way the search is performed. Most of the settings are hidden when you first go to edit the widget and you will need to click the show/hide config button to get access to many of the parameters.

**Label** sets the label displayed next to the field, you can set this to whatever you want

**Data Type** controls what type of data is being searched (i.e. what database table is being used). The standard options are "Post Field" for data such as post title and post content, from the standard wordpress post data, "Custom Field" for data from the custom fields, and "Category" to search on the categories a post is in.

**Data Field** makes a more specific selection from the data available. The drop down list should give some sensible options, or if you know the name of the database field you want to query you can type this in manually.

**Widget** controls what type of HTML input is created for the front end of the site.  It should be fairly obvious what these do.  

Some widgets will require extra options in the **Widget Options** field. At the moment this is just the drop down and radio button widgets which allow you to specify a list of values for the user to choose from. If you leave the options blank then they will be automatically populated with a list of all values currently in the database, this can be useful for fields like categories and tags, but is less useful for fields like title which would just generate an entry for every post in the blog. If you want to specify the values manually you can specify this as a comma separated list of values, "a,b,c" for three options with values "a", "b" and "c" respectively, or if you want to give "friendly" labels to the user you can separate the value from the label using a colon as follows "a:Group A,b:Group B,c:Group C".

Hidden constant fields allow you to add a fixed search parameter to the form which the user cannot edit.  The extra options allow you to specify what the value of this fixed parameter is.

The **Compare** field controls the way that the user input is compared to the data in the database. I would expect the most commonly used of these to be "equals" which requires an exact match between the user data and the database field (useful for category searches) and the "Words in" or "Phrase In" types which will search to see if the user input is a part of the data, rather than a full match, this is useful for things like searching the text of a post. The difference between "words" and "phrase" is that "words" splits the input into a series of words and searches for these individually whereas "Phrase" searches for all the words in sequence.

The "Less Than" and "More Than" comparisons ensure that the data is less than or more than the user input respectively. This is probably only relevant for numeric input but it can be used for strings as well and will search alphabetically. For numeric input it can help to also tick the "Numeric" checkbox as sometimes the database will be set up to compare numbers alphabetically (so that 10 is less than 2).

The "Range" comparison is a little more complicated and requires the user input to be of the form A-B to return results that are between A and B. This is primarily intended to be used for Drop-downs and radio buttons where the values can be set by the site administrator. If you specify the values for the dropdown like the following: "-10:Less Than 10,10-20:Between 10 and 20,20-:More than 20", then the user will be presented with the options "Less Than 10", "Betwen 10 and 20" and "More than 20" but the search will be done with the values "-10","10-20" and "20-". Again the numeric checkbox should be used for numeric data.

I hope that helps, please let me know if you have any trouble (or joy) with the plugin or if any of my instructions don"t make sense.

== Using Presets == 

Presets are a new feature in version 0.3, they allow search forms to be configured separately from the sidebar widgets, the configured form doesn"t directly appear on the site but can be included in two ways.

Firstly, by copying and pasting the php code from the preset config page into a template file you can place a search form anywhere you want in your template.

Secondly, by using a preset as the basis for another search form.  When setting up a search form you have the option of selecting one of the presets from a dropdown list.  Selecting one will include all of the fields from this preset in your form.

Presets are configured in exactly the same way as the sidebar widget, you can find the presets in the "Settings " WP Custom Fields Search" section of your admin site.


== Extending Custom Search == 

For developers, I"ve tried to build this in a modular way to allow extensions to be added, I will try to document this at some point in the future but if you can"t wait then the best places to get started would be to look at the filter "custom_search_get_classes" in custom-search/custom-search.php and the [Great Real Estate](http://www.rogertheriault.com/agents/plugins/great-real-estate-plugin/) bridging code in custom-search/bridges/greatrealestate.php.

This plugin is at quite an early stage in development though so it is likely that future versions may not be compatible with the same extension interface and it is likely that any extensions will need altering when new versions are released..



CHANGELOG.txt
2015-07-27	Release 0.3.27
	Removed dependence on mysql_query as this was failing in setups using mysqli
	Sets the is_home variable correctly
2013-10-07	Release 0.3.26
	Added drag and drop ordering of search fields.
2013-10-07	Release 0.3.25
	Removed Flexbox depency and replaced with jquery autocomplete
	Promotional link is now opt-in
2013-09-30	Release 0.3.24
	Now allows commas in dropdown options.
2013-06-27	Release 0.3.23
	Added widget wrapper
2013-06-27	Release 0.3.22
	Suppressed a bunch of warnings
	Removed a loose closing div
2013-06-27	Release 0.3.21
	Fixed bug causing incompatibility with relevanssi
2013-06-20	Release 0.3.20
	The checkbox functionality had broken the custom fields search.  This was a bit of an oversight on my part, sorry.
2013-06-17	Release 0.3.19
	There was a problem with the switch to using IDs for categories, this fix is not perfect but search was entirely broken for those affected before.  The problems are that there are now slightly blurred lines between keys and values in dropdown fields and category searches.
2013-06-03	Release 0.3.18
	Removed a warning in the sort functionality
	Changed branding
2013-06-03	Release 0.3.17
	Fixed bug with widgets and newer versions of jquery
	Added simple checkbox inputs
	Merged in changes from a number of bespoke versions
29/12/2010	Bespoke Custom Version
	Added result ordering and is_search compatibility.
19/08/2009	Release 0.3.16
	Fixed bug with radio buttons showing their config instead of options.
	Added some FAQs to the readme file.
11/08/2009	Release 0.3.15
	Fixed bug with custom fields joining.  Sorry everyone.
10/08/2009	Release 0.3.14
	Added all option for custom fields
	Corrected short code syntax.
	Added AtLeast and MoreThan joiners
10/07/2009	Release 0.3.13
	Added in post type joiner field
	Added in paypal donate link
	Added in Pierre's french translation
	Corrected James' bug with preset names
	Merged in alternative category joining
24/05/2009	Release 0.3.12
	Fixed flexbox integration for compatibility with newer version of jquery.
24/05/2009	Release 0.3.11
	Fixed typo in variable name which was preventing more than fields from working.
29/04/2009	Release 0.3.10
	Fixed incompatibility between real estate bridge and certain wordpress installs.
24/04/2009	Release 0.3.9
	Fixed Apache v1 config issue (Thanks to Petros for help in tracking this one down)
23/04/2009	Release 0.3.8
	Fixed IE config issue (Thanks to www.smesa.co.za for help in tracking this one down)
17/04/2009	Release 0.3.7
	Fixes a couple of show stopping bugs reporting by sheme
14/04/2009	Release 0.3.6
	Added hidden constants field type.
10/04/2009
	Sorts drop down options 
	Allows use of multiple category inputs
30/03/2009	Release 0.3.5
	Fixed short tags in templates (thanks to Dr North)
	Fixed bug in loading of search templates.
23/03/2009	Release 0.3.4
	Fixed bug whereby the global namespace was being polluted during the initialisation, preventing theme and plugin editing.
23/03/2009	Release 0.3.3
	Added tag syntax for placing search forms in posts.
23/03/2009	Release 0.3.2
	Added Wordpress.org format readme file.
23/03/2009	Release 0.3
	Added presets functionality
20/03/2009
	Works when run via a rewritten URL (in some cases anyway)
	Compatible with WP 2.7 and jQuery 1.2.3
18/03/2009 
	Drop down values with quotes are re-selected in the search form after search.
	Back end field suggestions fixed for PHP4
17/03/2009 
	Made a more sensible default form and hides the form parameters until clicked to produce a slightly less daunting interface
	Automatic drop downs only populated from live fields.
	Corrected a problem with defaults in PHP4 
	Always joins for certain search types, prevents un-typed content appearing in searches with no parameters/

12/03/2009	Initial Release 0.1


wp-custom-fields-search.php
<?php
/*
Plugin Name: WP Custom Search
Plugin URI: http://www.webhammer.co.uk/bespoke-wordpress-plugins/wp-custom-search
Description: Allows admin to build custom search form.  Allows the site admin to configure multiple html inputs for different fields including custom fields.  Also provides an extensible mechanism for integrating with other plugins data structures.
Version: 0.3.27
Author: Don Benjamin
Author URI: http://www.webhammer.co.uk/
Text Domain: wp-custom-fields-search
*/
/*
 * Copyright 2009-2013 Don Benjamin
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at 
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, 
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 * See the License for the specific language governing permissions and 
 * limitations under the License.
 */
	require_once(dirname(__FILE__).'/extra_search_fields.php');

	//Add Widget for configurable search.
	add_action('plugins_loaded',array('DB_CustomSearch_Widget','init'));


	class DB_CustomSearch_Widget extends DB_Search_Widget {
		function DB_CustomSearch_Widget($params=array()){
			DB_CustomSearch_Widget::__construct($params);
		}
		function __construct($params=array()){
			$this->loadTranslations();
			parent::__construct(__('Custom Fields ','wp-custom-fields-search'),$params);
			add_action('admin_print_scripts', array(&$this,'print_admin_scripts'), 90);
			add_action('admin_menu', array(&$this,'plugin_menu'), 90);
			add_filter('the_content', array(&$this,'process_tag'),9);
			add_shortcode( 'wp-custom-fields-search', array(&$this,'process_shortcode') );
			wp_enqueue_script('jquery');
			wp_enqueue_script('jquery-ui');
			wp_enqueue_script('jquery-ui-autocomplete');
			wp_enqueue_script('jquery-ui-sortable');
			if(version_compare("2.7",$GLOBALS['wp_version'])>0) wp_enqueue_script('dimensions');
		}
		function init(){
			global $CustomSearchFieldStatic;
			$CustomSearchFieldStatic['Object'] = new DB_CustomSearch_Widget();
			$CustomSearchFieldStatic['Object']->ensureUpToDate();
		}

		function currentVersion(){
			return "0.3.27";
		}

		function ensureUpToDate(){
			$version = $this->getConfig('version');
			$latest = $this->currentVersion();
			if($version<$latest) $this->upgrade($version,$latest);
		}

		function upgrade($current,$target){
			$options = $this->getConfig();
			if(version_compare($current,"0.3")<0){
				$config = $this->getDefaultConfig();
				$config['name'] = __('Default Preset','wp-custom-fields-search');
				$options['preset-default'] = $config;
			}
			$options['version']=$target;
			update_option($this->id,$options);
		}

		function getInputs($params = false,$visitedPresets=array()){
			if(is_array($params)){
				$id = $params['widget_id'];
			} else {
				$id = $params;
			}
			if(@$visitedPresets[$id]) return array();
			$visitedPresets[$id]=true;
			
			global $CustomSearchFieldStatic;
			if(!@$CustomSearchFieldStatic['Inputs'][$id]){
				$classes = $this->getClasses('joiner');
			
				$config = $this->getConfig($id);
				$inputs = array();
				if($config['preset']) $inputs = $this->getInputs($config['preset'],$visitedPresets);
				$nonFields = $this->getNonInputFields();
				if($config)
				foreach($config as $k=>$v){
					if(in_array($k,$nonFields)) continue;
					if(!(class_exists($v['input']) && class_exists($v['comparison']) && class_exists($v['joiner']))) {
						continue;
					}
					$inputs[] =  new CustomSearchField($v);

				}
				foreach($inputs as $k=>$v){
					$inputs[$k]->setIndex($k);
				}
				$CustomSearchFieldStatic['Inputs'][$id]=$inputs;
			}
			return $CustomSearchFieldStatic['Inputs'][$id];
		}
		function getTitle($params){
			$config = $this->getConfig($params['widget_id']);
			return $config['name'];
		}

		function form_processPost($post,$old){
			unset($post['###TEMPLATE_ID###']);
			if(!$post) $post=array('exists'=>1);
			$post = stripslashes_deep($post);
			return $post;
		}
		function getDefaultConfig(){
			return apply_filters('custom_search_default_config',
			array('name'=>'Site Search', 
				1=>array(
					'label'=>__('Key Words','wp-custom-fields-search'),
					'input'=>'TextField',
					'comparison'=>'WordsLikeComparison',
					'joiner'=>'PostDataJoiner',
					'name'=>'all'
				),
				2=>array(
					'label'=>__('Category','wp-custom-fields-search'),
					'input'=>'DropDownField',
					'comparison'=>'EqualComparison',
					'joiner'=>'CategoryJoiner'
				),
			));
		}
		function form_outputForm($values,$pref){
			$defaults=$this->getDefaultConfig();
			$prefId = preg_replace('/^.*\[([^]]*)\]$/','\\1',$pref);
			$this->form_existsInput($pref);
			$rand = rand();
?>
	<div id='config-template-<?php echo $prefId?>' style='display: none;'>
	<?php 
		$templateDefaults = $defaults[1];
		$templateDefaults['label'] = 'Field ###TEMPLATE_ID###';
		echo  $this->singleFieldHTML($pref,'###TEMPLATE_ID###',$templateDefaults);
	?>
	</div>

<?php
			foreach($this->getClasses('input') as $class=>$desc) {
				if(class_exists($class))
					$form = new $class();
				else $form = false;
				if(compat_method_exists($form,'getConfigForm')){
					if($form = $form->getConfigForm($pref.'[###TEMPLATE_ID###]',array('name'=>'###TEMPLATE_NAME###'))){
?>
	<div id='config-input-templates-<?php echo $class?>-<?php echo $prefId?>' style='display: none;'>
		<?php echo $form?>
	</div>
		
<?php					}
				}
			}
 ?>
	<div id='config-form-<?php echo $prefId?>'>
<?php
			if(!$values) $values = $defaults;
			$maxId=0;
			$presets = $this->getPresets();
			array_unshift($presets,__('NONE','wp-custom-fields-search'));
?>
		<div class='searchform-name-wrapper'><label for='<?php echo $prefId?>[name]'><?php echo __('Search Title','wp-custom-fields-search')?></label><input type='text' class='form-title-input' id='<?php echo $prefId?>[name]' name='<?php echo $pref?>[name]' value='<?php echo @$values['name']?>'/></div>
		<div class='searchform-preset-wrapper'><label for='<?php echo $prefId?>[preset]'><?php echo __('Use Preset','wp-custom-fields-search')?></label>
<?php
			$dd = new AdminDropDown($pref."[preset]",@$values['preset'],$presets);
			echo $dd->getInput()."</div>";
?>
	<div class='searchform-link-wrapper'>
		<input type='checkbox' id='<?php echo $pref?>[showlink]'
		 name='<?php echo $pref?>[showlink]' <?php if(@$values['showlink']) { ?>checked='checked'<?php } ?> />
		<label for='<?php echo $pref?>[showlink]'><?php echo __('Support plugin by including link to author\'s site','wp-custom-fields-search')?></label>
	</div>
	<div class='searchform-fields-wrapper sort-wrapper'>
<?php
			$nonFields = $this->getNonInputFields();
			foreach($values as $id => $val){
				$maxId = max($id,$maxId);
				if(in_array($id,$nonFields)) continue;
				echo "<div id='config-form-$prefId-$id' class='sort-element'>".$this->singleFieldHTML($pref,$id,$val)."</div>";
			}
?>
	</div>
	</div>

	<br/><a href='#' onClick="return CustomSearch.get('<?php echo $prefId?>').add();"><?php echo __('Add Field','wp-custom-fields-search')?></a>
	<script type='text/javascript'>
		jQuery(function(){
			CustomSearch.create('<?php echo $prefId?>','<?php echo $maxId?>');
<?php
	foreach($this->getClasses('joiner') as $joinerClass=>$desc){
		if(compat_method_exists($joinerClass,'getSuggestedFields')){
			$options = eval("return $joinerClass::getSuggestedFields();");
			$str = '';
			if(is_array($options)){
				foreach($options as $i=>$v){
					$k=$i;
					if(is_numeric($k)) $k=$v;
					$options[$i] = json_encode(array('id'=>$k,'name'=>$v));
				}
				$str = '['.join(',',$options).']';
			} elseif(is_string($options)) {
				$str = json_encode($options);
			}
			echo "CustomSearch.setOptionsFor('$joinerClass',".$str.");\n";
		}elseif(eval("return $joinerClass::needsField();")){
			echo "CustomSearch.setOptionsFor('$joinerClass','any');\n";
		}
	}
?>
	});
	</script>
<?php
		}

		function getNonInputFields(){
			return array('exists','name','preset','version','order','template','showlink');
		}
		function singleFieldHTML($pref,$id,$values){
			$prefId = preg_replace('/^.*\[([^]]*)\]$/','\\1',$pref);
			$pref = $pref."[$id]";
			$htmlId = $pref."[exists]";
			$output = "<input type='hidden' name='$htmlId' value='1'/>";
			$titles="<th>".__('Label','wp-custom-fields-search')."</th>";
			$inputs="<td><input type='text' name='$pref"."[label]' value='$values[label]' class='form-field-title'/></td><td><a href='#' onClick='return CustomSearch.get(\"$prefId\").toggleOptions(\"$id\");'>".__('Show/Hide Config','wp-custom-fields-search')."</a></td>";
			$output.="<table class='form-field-table'><tr>$titles</tr><tr>$inputs</tr></table>";
			$output.="<div id='form-field-advancedoptions-$prefId-$id' style='display: none'>";
			$inputs='';$titles='';
			$titles="<th>".__('Data Field','wp-custom-fields-search')."</th>";
			$inputs="<td><div id='form-field-dbname-$prefId-$id' class='form-field-title-div'><input type='text' name='$pref"."[name]' value='".@$values["name"]."' class='form-field-title'/></div></td>";
			$count=1;
			foreach(array('joiner'=>__('Data Type','wp-custom-fields-search'),'comparison'=>__('Compare','wp-custom-fields-search'),'input'=>__('Widget','wp-custom-fields-search')) as $k=>$v){
				$dd = new AdminDropDown($pref."[$k]",$values[$k],$this->getClasses($k),array('onChange'=>'CustomSearch.get("'.$prefId.'").updateOptions("'.$id.'","'.$k.'")','css_class'=>"wpcfs-$k"));
				$titles="<th>".$v."</th>".$titles;
				$inputs="<td>".$dd->getInput()."</td>".$inputs;
				if(++$count==2){
					$output.="<table class='form-field-table form-class-$k'><tr>$titles</tr><tr>$inputs</tr></table>";
					$count=0;
					$inputs = $titles='';
				}
			}
			if($titles){
				$output.="<table class='form-field-table'><tr>$titles</tr><tr>$inputs</tr></table>";
				$inputs = $titles='';
			}
			$titles.="<th>".__('Numeric','wp-custom-fields-search')."</th><th>".__('Widget Config','wp-custom-fields-search')."</th>";
			$inputs.="<td><input type='checkbox' ".(@$values['numeric']?"checked='true'":"")." name='$pref"."[numeric]'/></td>";

			$widgetConfig='';
			if(class_exists($widgetClass = $values['input'])){
				$widget = new $widgetClass();
				if(compat_method_exists($widget,'getConfigForm'))
					$widgetConfig=$widget->getConfigForm($pref,$values);
			}


			$inputs.="<td><div id='$this->id"."-$prefId"."-$id"."-widget-config'>$widgetConfig</div></td>";
			$output.="<table class='form-field-table'><tr>$titles</tr><tr>$inputs</tr></table>";
			$output.="</div>";
			$output.="<a href='#' onClick=\"return CustomSearch.get('$prefId').remove('$id');\">Remove Field</a>";
			return "<div class='field-wrapper'>$output</div>";
		}

		function getRootURL(){
			return WP_CONTENT_URL .'/plugins/' .  dirname(plugin_basename(__FILE__) ) . '/';
		}
		function print_admin_scripts($params){
			$jsRoot = $this->getRootURL().'js';
			$cssRoot = $this->getRootURL().'css';
			$scripts = array('Class.js','CustomSearch.js');
			foreach($scripts as $file){
				echo "<script src='$jsRoot/$file' ></script>";
			}
			echo "<link rel='stylesheet' href='$cssRoot/admin.css' >";
		}

		function getJoiners(){
			return $this->getClasses('joiner');
		}
		function getComparisons(){
			return $this->getClasses('comparison');
		}
		function getInputTypes(){
			return $this->getClasses('input');
		}
		function getClasses($type){
			global $CustomSearchFieldStatic;
			if(!$CustomSearchFieldStatic['Types']){
				$CustomSearchFieldStatic['Types'] = array(
					"joiner"=>array(
						"PostDataJoiner" =>__( "Post Field",'wp-custom-fields-search'),
						"CustomFieldJoiner" =>__( "Custom Field",'wp-custom-fields-search'),
						"CategoryJoiner" =>__( "Category",'wp-custom-fields-search'),
						"TagJoiner" =>__( "Tag",'wp-custom-fields-search'),
						"PostTypeJoiner" =>__( "Post Type",'wp-custom-fields-search'),
					),
					"input"=>array(
						"TextField" =>__( "Text Input",'wp-custom-fields-search'),
						"DropDownField" =>__( "Drop Down",'wp-custom-fields-search'),
						"RadioButtonField" =>__( "Radio Button",'wp-custom-fields-search'),
						"HiddenField" =>__( "Hidden Constant",'wp-custom-fields-search'),
					),
					"comparison"=>array(
						"EqualComparison" =>__( "Equals",'wp-custom-fields-search'),
						"LikeComparison" =>__( "Phrase In",'wp-custom-fields-search'),
						"WordsLikeComparison" =>__( "Words In",'wp-custom-fields-search'),
						"LessThanComparison" =>__( "Less Than",'wp-custom-fields-search'),
						"MoreThanComparison" =>__( "More Than",'wp-custom-fields-search'),
						"AtMostComparison" =>__( "At Most",'wp-custom-fields-search'),
						"AtLeastComparison" =>__( "At Least",'wp-custom-fields-search'),
						"RangeComparison" =>__( "Range",'wp-custom-fields-search'),
//TODO: Make this work...
//						"NotEqualComparison" =>__( "Not Equal To",'wp-custom-fields-search'),
					)
				);
				$CustomSearchFieldStatic['Types'] = apply_filters('custom_search_get_classes',$CustomSearchFieldStatic['Types']);
			}
			return $CustomSearchFieldStatic['Types'][$type];
		}
		function plugin_menu(){
			add_options_page('Form Presets','WP Custom Fields Search',8,__FILE__,array(&$this,'presets_form'));
		}
		function getPresets(){
			$presets = array();
			foreach(array_keys($config = $this->getConfig()) as $key){
				if(strpos($key,'preset-')===0) {
					$presets[$key] = $key;
					if($name = $config[$key]['name'])
						$presets[$key]=$name;
				}
			}
			return $presets;
		}
		function presets_form(){
			$presets=$this->getPresets();
			if(!$preset = @$_REQUEST['selected-preset']){
				$preset = 'preset-default';
			}
			if(!$presets[$preset]){
				$defaults = $this->getDefaultConfig();
				$options = $this->getConfig();
				$options[$preset] = $defaults;
				if($n = $_POST[$this->id][$preset]['name'])
					$options[$preset]['name'] = $n;
				elseif($preset=='preset-default')
					$options[$preset]['name'] = 'Default';
				else{
					list($junk,$id) = explode("-",$preset);
					$options[$preset]['name'] = 'New Preset '.$id;
				}
				update_option($this->id,$options);
				$presets[$preset] = $options[$preset]['name'];
			}
			if(@$_POST['delete']){
				check_admin_referer($this->id.'-editpreset-'.$preset);
				$options = $this->getConfig();
				unset($options[$preset]);
				unset($presets[$preset]);
				update_option($this->id,$options);
				list($preset,$name) = each($presets);
			}

			$index = 1;
			while(@$presets["preset-$index"]) $index++;
			$presets["preset-$index"] = __('New Preset','wp-custom-fields-search');

			$linkBase = $_SERVER['REQUEST_URI'];
			$linkBase = preg_replace("/&?selected-preset=[^&]*(&|$)/",'',$linkBase);
			foreach($presets as $key=>$name){
				$config = $this->getConfig($key);
				if($config && $config['name']) $name=$config['name'];
				if(($n = @$_POST[$this->id][$key]['name'])&&(!@$_POST['delete']))
					$name = $n;
				$presets[$key]=$name;
			}
			$plugin=&$this;
			ob_start();
			wp_nonce_field($this->id.'-editpreset-'.$preset);
			$hidden = ob_get_contents();
			$hidden.="<input type='hidden' name='selected-preset' value='$preset'>";
			$shouldSave = $_POST['selected-preset'] && !$_POST['delete'] && check_admin_referer($this->id.'-editpreset-'.$preset);
			ob_end_clean();
			include(dirname(__FILE__).'/templates/options.php');
		}
		function process_tag($content){
			$regex = '/\[\s*wp-custom-fields-search\s+(?:([^\]=]+(?:\s+.*)?))?\]/';
			return preg_replace_callback($regex, array(&$this, 'generate_from_tag'), $content);
		}
		function process_shortcode($atts,$content){
			return $this->generate_from_tag(array("",$atts['preset']));
		}
		function generate_from_tag($reMatches){
			global $CustomSearchFieldStatic;
			ob_start();

			$preset=$reMatches[1];
			if(!$preset) $preset = 'default';
			wp_custom_fields_search($preset);

			$form = ob_get_contents();
			ob_end_clean();
			return $form;
		}
	}
	global $CustomSearchFieldStatic;
	$CustomSearchFieldStatic['Inputs'] = array();
	$CustomSearchFieldStatic['Types'] = array();

	class AdminDropDown extends DropDownField {
		function AdminDropDown($name,$value,$options,$params=array()){
			AdminDropDown::__construct($name,$value,$options,$params);
		}
		function __construct($name,$value,$options,$params=array()){
			$params['options'] = $options;
			$params['id'] = @$params['name'];
			parent::__construct($params);
			$this->name = $name;
			$this->value = $value;
		}
		function getHTMLName(){
			return $this->name;
		}
		function getValue(){
			return $this->value;
		}
		function getInput(){
			return parent::getInput($this->name,null);
		}
	}

if (!function_exists('json_encode'))
{
  function json_encode($a=false)
  {
    if (is_null($a)) return 'null';
    if ($a === false) return 'false';
    if ($a === true) return 'true';
    if (is_scalar($a))
    {
      if (is_float($a))
      {
        // Always use "." for floats.
        return floatval(str_replace(",", ".", strval($a)));
      }

      if (is_string($a))
      {
        static $jsonReplaces = array(array("\\", "/", "\n", "\t", "\r", "\b", "\f", '"'), array('\\\\', '\\/', '\\n', '\\t', '\\r', '\\b', '\\f', '\"'));
        return '"' . str_replace($jsonReplaces[0], $jsonReplaces[1], $a) . '"';
      }
      else
        return $a;
    }
    $isList = true;
    for ($i = 0, reset($a); $i < count($a); $i++, next($a))
    {
      if (key($a) !== $i)
      {
        $isList = false;
        break;
      }
    }
    $result = array();
    if ($isList)
    {
      foreach ($a as $v) $result[] = json_encode($v);
      return '[' . join(',', $result) . ']';
    }
    else
    {
      foreach ($a as $k => $v) $result[] = json_encode($k).':'.json_encode($v);
      return '{' . join(',', $result) . '}';
    }
  }
}
function wp_custom_fields_search($presetName='default'){
	global $CustomSearchFieldStatic;
	if(strpos($presetName,'preset-')!==0) $presetName="preset-$presetName";
	$CustomSearchFieldStatic['Object']->renderWidget(array('widget_id'=>$presetName,'noTitle'=>true),array('number'=>$presetName));
}
function compat_method_exists($class,$method){
	return method_exists($class,$method) || in_array(strtolower($method),get_class_methods($class));
}
